pretyping/evarsolve.ml

```diff
@@ -1455,7 +1464,11 @@ let occur_evar_upto_types sigma n c =
 let instantiate_evar unify flags env evd evk body =
   (* Check instance freezing the evar to be defined, as
      checking could involve the same evar definition problem again otherwise *)
-  let flags = { flags with frozen_evars = Evar.Set.add evk flags.frozen_evars } in
+  let allowed = match flags.allowed_evars with
+    | AllowAll -> fun evk' -> not (Evar.equal evk evk')
+    | AllowFun f -> fun evk' -> not (Evar.equal evk evk') && f evk'
```

Comment:
This is going to be O(n), so potentially slow...

Reply:
I know, but what should I do?

Reply:
Offline discussion with @ppedrot: use a datastructure based on a map + a set in both unification engines.

Reply:
Done.